### PR TITLE
koordlet: fix failed update existing Device CRD object

### LINKED ISSUE #629: [BUG] Device CRD is empty when node does have GPU
**What happened**:
Node `cn-hangzhou.10.0.2.216` has 4 V100s mounted. 
```
[root@iZ8Z ~]# nvidia-smi
Fri Sep 16 16:37:45 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.91.03    Driver Version: 460.91.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:00:07.0 Off |                    0 |
| N/A   34C    P0    41W / 300W |      0MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:00:08.0 Off |                    0 |
| N/A   34C    P0    41W / 300W |      0MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:00:09.0 Off |                    0 |
| N/A   34C    P0    39W / 300W |      0MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  On   | 00000000:00:0A.0 Off |                    0 |
| N/A   35C    P0    41W / 300W |      0MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

However, Device CRD is empty.
```
% kubectl get devices cn-hangzhou.10.0.2.216 -o yaml
apiVersion: scheduling.koordinator.sh/v1alpha1
kind: Device
metadata:
  creationTimestamp: "2022-09-16T08:31:06Z"
  generation: 1
  name: cn-hangzhou.10.0.2.216
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Node
    name: cn-hangzhou.10.0.2.216
    uid: 2b0e0cd1-05bb-4238-b6df-b98e54e1f998
  resourceVersion: "8845655"
  uid: 919d643f-18eb-499e-96ed-eea680237374
spec: {}
status: {}
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- App version: commit `15d801f`
- Kubernetes version (use `kubectl version`):
- Install details (e.g. helm install args):
- Others:

### PR DESCRIPTION
**Pull request by eahydra**
Signed-off-by: Joseph <joseph.t.lee@outlook.com>

### Ⅰ. Describe what this PR does

1. fix update Device CRD object failed  if koordlet restart or refresh
2. if the spec of device has not changed, ignore update.

### Ⅱ. Does this pull request fix one issue?
fixes #629

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/statesinformer/states_device_linux.go [hunk 1]
```diff
@@ -66,17 +67,34 @@ func (s *statesInformer) reportDevice() {
...
+		return
+	}
+	if !errors.IsAlreadyExists(err) {
+		klog.Errorf("Failed to create Device %s, err: %v", node.Name, err)
+		return
+	}
+
+	klog.V(4).Infof("Device %s already exists, try to update with latest Devices", node.Name)
+	err = util.RetryOnConflictOrTooManyRequests(func() error {
+		device, err := s.deviceClient.Get(context.TODO(), device.Name, metav1.GetOptions{ResourceVersion: "0"})
```
Comment (WeizhouRen) [inline pkg/koordlet/statesinformer/states_device_linux.go:82]: do we need `ResourceVersion: "0"` when calling `Get` function?

File: pkg/koordlet/statesinformer/states_device_linux.go [hunk 2]
```diff
@@ -66,17 +67,34 @@ func (s *statesInformer) reportDevice() {
...
+		device.Spec = schedulingv1alpha1.DeviceSpec{
+			Devices: gpuDevices,
 		}
+		_, err = s.deviceClient.Update(context.TODO(), device, metav1.UpdateOptions{})
+		return err
+	})
+	if err != nil {
+		klog.Errorf("failed to update device %s, err: %v", node.Name, err)
+	} else {
+		klog.V(4).Infof("successfully or update Device %s", device.Name)
```
Comment (WeizhouRen) [inline pkg/koordlet/statesinformer/states_device_linux.go:97]: successfully or? I guess it's a typo.

File: pkg/koordlet/statesinformer/states_device_linux.go [hunk 3]
```diff
@@ -66,17 +67,34 @@ func (s *statesInformer) reportDevice() {
...
+	if err == nil {
+		klog.V(4).Infof("successfully create Device %s", device.Name)
+		return
+	}
+	if !errors.IsAlreadyExists(err) {
+		klog.Errorf("Failed to create Device %s, err: %v", node.Name, err)
+		return
+	}
+
+	klog.V(4).Infof("Device %s already exists, try to update with latest Devices", node.Name)
```
Comment (WeizhouRen) [inline pkg/koordlet/statesinformer/states_device_linux.go:80]: level 5 would be better, WDYT?

Comment (eahydra) [reply pkg/koordlet/statesinformer/states_device_linux.go:80]: We need the log.

File: pkg/koordlet/statesinformer/states_device_linux.go (hunk 2 above)
Comment (eahydra) [reply pkg/koordlet/statesinformer/states_device_linux.go]: fixed

**Review by jasonliu747 — COMMENTED** [review]
/lgtm

**Pull request merged**